src/Microsoft.AspNetCore.Razor.Tools/ConcurrentLruCache.cs

```diff
@@ -0,0 +1,238 @@
+        /// <see cref="ArgumentException"/> will be thrown.
+        /// </summary>
+        public ConcurrentLruCache(KeyValuePair<K, V>[] array)
+            : this(array.Length)
+        {
+            foreach (var kvp in array)
+            {
+                this.UnsafeAdd(kvp.Key, kvp.Value, true);
+            }
+        }
+
+        /// <summary>
+        /// For testing. Very expensive.
+        /// </summary>
+        internal IEnumerable<KeyValuePair<K, V>> TestingEnumerable
+        {
+            get
+            {
+                lock (_lockObject)
+                {
+                    var copy = new KeyValuePair<K, V>[_cache.Count];
+                    int index = 0;
+                    foreach (K key in _nodeList)
+                    {
+                        copy[index++] = new KeyValuePair<K, V>(key,
+                                                               _cache[key].Value);
+                    }
+                    return copy;
+                }
+            }
+        }
+
+        public void Add(K key, V value)
+        {
+            lock (_lockObject)
+            {
+                UnsafeAdd(key, value, true);
+            }
+        }
+
+        private void MoveNodeToTop(LinkedListNode<K> node)
+        {
+            if (!object.ReferenceEquals(_nodeList.First, node))
+            {
+                _nodeList.Remove(node);
+                _nodeList.AddFirst(node);
+            }
+        }
+
+        /// <summary>
+        /// Expects non-empty cache. Does not lock.
+        /// </summary>
+        private void UnsafeEvictLastNode()
+        {
+            Debug.Assert(_capacity > 0);
+            var lastNode = _nodeList.Last;
+            _nodeList.Remove(lastNode);
+            _cache.Remove(lastNode.Value);
+        }
+
+        private void UnsafeAddNodeToTop(K key, V value)
+        {
+            var node = new LinkedListNode<K>(key);
+            _cache.Add(key, new CacheValue { Node = node, Value = value });
+            _nodeList.AddFirst(node);
+        }
+
+        /// <summary>
+        /// Doesn't lock.
+        /// </summary>
+        private void UnsafeAdd(K key, V value, bool throwExceptionIfKeyExists)
+        {
+            if (_cache.TryGetValue(key, out var result))
+            {
+                if (throwExceptionIfKeyExists)
+                {
+                    throw new ArgumentException("Key already exists", nameof(key));
+                }
+                else if (!result.Value.Equals(value))
+                {
+                    result.Value = value;
+                    _cache[key] = result;
+                    MoveNodeToTop(result.Node);
+                }
+            }
+            else
+            {
+                if (_cache.Count == _capacity)
+                {
+                    UnsafeEvictLastNode();
+                }
+                UnsafeAddNodeToTop(key, value);
+            }
+        }
+
+        public V this[K key]
```

Comment:
It doesn't look like we use this method from the caller - all I saw was a `GetOrAdd`. Could we remove unused code - reduces the burden of testing \ maintaining the type?

Reply:
Using this in tests now. Removed other unused methods though.